File: app/src/main/java/yiyo/gitlabandroid/model/rest/RestClient.java
package yiyo.gitlabandroid.model.rest;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import retrofit.RestAdapter;
import retrofit.converter.GsonConverter;

/**
 * Created by yiyo on 12/07/15.
 */
public class RestClient {

    private static final String BASE_ULR = "https://gitlab.com/api/v3";
    private static ApiService apiService;
    private static RestClient INSTANCE;

    public RestClient() {
        Gson gson = new GsonBuilder()
                .setDateFormat("yyyy'-'MM'-'dd'T'HH':'mm':'ss'.'SSS'Z'")
                .create();

        RestAdapter restAdapter = new RestAdapter.Builder()
                .setLogLevel(RestAdapter.LogLevel.FULL)
                .setEndpoint(BASE_ULR)
                .setConverter(new GsonConverter(gson))
                .build();

        apiService = restAdapter.create(ApiService.class);
    }

    public static ApiService getApiService() {
        if (INSTANCE == null) {
            INSTANCE = new RestClient();
        }
        return apiService;
    }
}

File: app/src/main/kotlin/yiyo/gitlabandroid/model/rest/RestClient.kt
package yiyo.gitlabandroid.model.rest

import android.content.Context
import com.google.gson.GsonBuilder
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import yiyo.gitlabandroid.utils.Configuration

/**
 * Created by yiyo
 */
class RestClient {

    companion object {
        val BASE_ULR = "https://gitlab.com/api/v3/"

        fun getApiService(context: Context): ApiService {
            val httpClient = OkHttpClient.Builder()
            httpClient.addInterceptor { chain ->
                val configuration = Configuration(context);
                val originalRequest = chain.request()
                val request = originalRequest.newBuilder()
                        .addHeader("PRIVATE-TOKEN", configuration.getPrivateToken())
                        .addHeader("Content-Type", "application/json")
                .method(originalRequest.method(), originalRequest.body())
                .build()

                chain.proceed(request)
            }
            val client = httpClient.build()
            val gson = GsonBuilder().setDateFormat("yyyy'-'MM'-'dd'T'HH':'mm':'ss'.'SSS'Z'").create()

            val restAdapter = Retrofit.Builder()
                    .baseUrl(BASE_ULR)
                    .addConverterFactory(GsonConverterFactory.create(gson))
                    .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                    .client(client)
                    .build()

            return restAdapter.create(ApiService::class.java)
        }
    }
}
